File: src/lib/routing.ts
/**
 * Shared OSM bike routing: turns a list of waypoints into a real cycling
 * path via public routers, entirely client side — no keys required. Used by
 * the route planner (multi-waypoint, profile-aware) and by the rejoin helper
 * (two points, speed matters more than profile choice).
 */
import { haversine } from "./gpx";
import { osrmDirection, type RideCue } from "./cues";

export type LatLon = {
  lat: number;
  lon: number;
  /** elevation in meters, when the router's geometry included one (BRouter only — OSRM doesn't) */
  ele?: number;
};

/** BRouter profile names, as accepted by the public/self-hosted BRouter server. */
export type BikeProfile = "trekking" | "fastbike" | "safety";

export const BIKE_PROFILES: { value: BikeProfile; label: string; description: string }[] = [
  { value: "trekking", label: "Trekking", description: "Balanced, all-around cycling route" },
  { value: "fastbike", label: "Fast", description: "Prefers direct roads for speed" },
  { value: "safety", label: "Safety", description: "Prefers quieter, bike-friendly ways" },
];

const OSRM_BIKE = "https://routing.openstreetmap.de/routed-bike/route/v1/bike";
const BROUTER_DEFAULT = "https://brouter.de/brouter";

/**
 * Self-hosters can point this at their own BRouter (or BRouter-compatible)
 * instance via VITE_BROUTER_URL — see .env.example. Defaults to the public
 * brouter.de server, same as before this was configurable.
 */
function brouterBaseUrl(): string {
  const configured = (import.meta.env.VITE_BROUTER_URL as string | undefined)?.trim();
  return configured && configured.length > 0 ? configured.replace(/\/$/, "") : BROUTER_DEFAULT;
}

export function pathLengthM(path: LatLon[]): number {
  let total = 0;
  for (let i = 1; i < path.length; i += 1) {
    total += haversine(path[i - 1].lat, path[i - 1].lon, path[i].lat, path[i].lon);
  }
  return total;
}

/**
 * Converts GeoJSON `[lon, lat]`/`[lon, lat, ele]` coordinate tuples to our
 * point shape, keeping a 3rd (elevation) value when the router included one
 * — BRouter's geojson track coordinates carry elevation this way; OSRM's
 * don't, so its paths just come out with `ele` unset, same as before.
 */
function toPath(coords: number[][]): LatLon[] {
  return coords.map(([lon, lat, ele]) => ({
    lat,
    lon,
    ...(Number.isFinite(ele) ? { ele } : {}),
  }));
}

/**
 * Walks an OSRM route's legs/steps into a flat cue list with cumulative
 * distance offsets, so multi-waypoint routes get one continuous instruction
 * list instead of restarting at zero for every leg. A step's `maneuver`
 * happens at the start of that step, so the running offset *before* adding
 * the step's own distance is exactly where the instruction applies.
 */
function parseOsrmSteps(route: {
  legs?: {
    steps?: { distance?: number; name?: string; maneuver?: { type?: string; modifier?: string } }[];
  }[];
}): RideCue[] {
  const cues: RideCue[] = [];
  let offsetM = 0;
  for (const leg of route.legs ?? []) {
    for (const step of leg.steps ?? []) {
      if (step.maneuver) {
        cues.push({
          atM: offsetM,
          direction: osrmDirection(step.maneuver.type ?? "", step.maneuver.modifier ?? null),
          name: step.name?.trim() || null,
        });
      }
      offsetM += Number(step.distance) || 0;
    }
  }
  return cues;
}

/** Routes through every waypoint in order via the public OSRM bike server, capturing turn-by-turn step instructions along the way. */
export async function fetchOsrmRoute(
  points: LatLon[],
  signal: AbortSignal,
): Promise<{ path: LatLon[]; distanceM: number; cues: RideCue[] }> {
  const coords = points.map((p) => `${p.lon},${p.lat}`).join(";");
  const response = await fetch(
    `${OSRM_BIKE}/${coords}?overview=full&geometries=geojson&steps=true`,
    { signal },
  );
  if (!response.ok) throw new Error(`OSRM ${response.status}`);
  const data = await response.json();
  const route = data?.routes?.[0];
  const coordinates = route?.geometry?.coordinates;
  if (!Array.isArray(coordinates) || coordinates.length < 2) throw new Error("OSRM: no geometry");
  return {
    path: toPath(coordinates),
    distanceM: Number(route.distance) || 0,
    cues: parseOsrmSteps(route),
  };
}

/**
 * Routes through every waypoint in order via BRouter, respecting the chosen
 * cycling profile. Doesn't return `cues`: BRouter's public server has no
 * documented, stable step-instruction format to parse (unlike OSRM's
 * `steps=true`), so a route that comes back from here falls back to
 * geometry-detected turns for its cue sheet (see `buildCueSheet` in
 * cues.ts) instead of risking a silently-wrong parse of an undocumented
 * field.
 */
export async function fetchBrouterRoute(
  points: LatLon[],
  profile: BikeProfile,
  signal: AbortSignal,
): Promise<{ path: LatLon[]; distanceM: number; cues: RideCue[] }> {
  const lonlats = points.map((p) => `${p.lon},${p.lat}`).join("|");
  const url = `${brouterBaseUrl()}?lonlats=${lonlats}&profile=${profile}&alternativeidx=0&format=geojson`;
  const response = await fetch(url, { signal });
  if (!response.ok) throw new Error(`BRouter ${response.status}`);
  const data = await response.json();
  const coordinates = data?.features?.[0]?.geometry?.coordinates;
  if (!Array.isArray(coordinates) || coordinates.length < 2)
    throw new Error("BRouter: no geometry");
  const path = toPath(coordinates);
  const declared = Number(data.features[0]?.properties?.["track-length"]);
  return { path, distanceM: Number.isFinite(declared) ? declared : pathLengthM(path), cues: [] };
}

export type RoutedPath = {
  /** the path through all waypoints, following real ways when possible */
  path: LatLon[];
  /** length of that path in metres */
  distanceM: number;
  /** true when the geometry comes from a cycling router, false for the straight-line fallback */
  routed: boolean;
  /** turn-by-turn step instructions the router provided, if any — see fetchOsrmRoute/fetchBrouterRoute */
  cues: RideCue[];
};

function straightLineFallback(points: LatLon[]): RoutedPath {
  return { path: points, distanceM: pathLengthM(points), routed: false, cues: [] };
}

/**
 * Bike-friendly route through the given waypoints (in order), trying one
 * router then the other. Falls back to straight lines between waypoints
 * when both routers are unreachable (offline, remote areas, self-hosted
 * BRouter misconfigured).
 *
 * `preferBrouter` puts BRouter first so the requested `profile` is actually
 * honored — OSRM's public bike profile is fixed and ignores it, so it only
 * makes a good *fallback*, not a good first choice when profile matters.
 */
export async function fetchRoute(
  points: LatLon[],
  {
    profile = "trekking",
    preferBrouter = false,
    signal,
  }: { profile?: BikeProfile; preferBrouter?: boolean; signal: AbortSignal },
): Promise<RoutedPath> {
  if (points.length < 2) return straightLineFallback(points);

  const attempts = preferBrouter
    ? [() => fetchBrouterRoute(points, profile, signal), () => fetchOsrmRoute(points, signal)]
    : [() => fetchOsrmRoute(points, signal), () => fetchBrouterRoute(points, profile, signal)];

  for (const attempt of attempts) {
    try {
      const result = await attempt();
      return { ...result, routed: true };
    } catch (error) {
      if (signal.aborted) throw error;
    }
  }
  return straightLineFallback(points);
}

File: android-native/app/src/main/java/app/hodora/mobile/routing/Routing.kt
package app.hodora.mobile.routing

import app.hodora.mobile.cues.RideCue
import app.hodora.mobile.cues.osrmDirection
import app.hodora.mobile.gpx.Bounds
import app.hodora.mobile.gpx.RidePoint
import app.hodora.mobile.gpx.haversine
import app.hodora.mobile.net.HttpClientProvider
import io.ktor.client.request.get
import io.ktor.client.statement.bodyAsText
import io.ktor.http.isSuccess
import kotlinx.coroutines.CancellationException
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * Direct port of src/lib/routing.ts: turns a list of waypoints into a real
 * cycling path via public OSM routers, client-side, no keys required. Same
 * endpoints, same fallback order, same straight-line-if-both-fail behavior.
 */
@Serializable
data class LatLon(
    val lat: Double,
    val lon: Double,
    /** elevation in meters, when the router's geometry included one (BRouter only — OSRM doesn't) */
    val ele: Double? = null,
)

/** BRouter profile names, as accepted by the public/self-hosted BRouter server. */
@Serializable
enum class BikeProfile(val apiValue: String, val label: String, val description: String) {
    @SerialName("trekking")
    TREKKING("trekking", "Trekking", "Balanced, all-around cycling route"),

    @SerialName("fastbike")
    FASTBIKE("fastbike", "Fast", "Prefers direct roads for speed"),

    @SerialName("safety")
    SAFETY("safety", "Safety", "Prefers quieter, bike-friendly ways"),
}

data class RoutedPath(
    /** the path through all waypoints, following real ways when possible */
    val path: List<LatLon>,
    /** length of that path in meters */
    val distanceM: Double,
    /** true when the geometry comes from a cycling router, false for the straight-line fallback */
    val routed: Boolean,
    /** turn-by-turn step instructions the router provided, if any */
    val cues: List<RideCue>,
)

private const val OSRM_BIKE = "https://routing.openstreetmap.de/routed-bike/route/v1/bike"

// Same zero-config public default as the web app (VITE_BROUTER_URL lets
// self-hosters override it there); not yet plumbed through to a settings
// screen here.
private const val BROUTER_DEFAULT = "https://brouter.de/brouter"

private val json = Json { ignoreUnknownKeys = true; isLenient = true }

fun pathLengthM(path: List<LatLon>): Double {
    var total = 0.0
    for (i in 1 until path.size) {
        total += haversine(path[i - 1].lat, path[i - 1].lon, path[i].lat, path[i].lon)
    }
    return total
}

private fun toLatLonPath(coordinates: List<List<Double>>): List<LatLon> =
    coordinates.map { c -> LatLon(lat = c[1], lon = c[0], ele = c.getOrNull(2)) }

/**
 * Converts a routed path back into a ride's point list — port of
 * toRidePoints in src/lib/discover.ts (used by both the route planner and
 * the loop generator on web; only the planner is ported so far). Unlike
 * gpx.ts's buildParsedRide, the web version leaves `d` as an unrounded
 * float; rounding it to match RidePoint.d's Int type here only costs
 * sub-meter precision.
 */
fun toRidePoints(path: List<LatLon>): List<RidePoint> {
    var d = 0.0
    return path.mapIndexed { index, point ->
        if (index > 0) {
            d += haversine(path[index - 1].lat, path[index - 1].lon, point.lat, point.lon)
        }
        RidePoint(lat = point.lat, lon = point.lon, ele = point.ele ?: 0.0, d = Math.round(d).toInt())
    }
}

/** Port of boundsOf in src/lib/discover.ts. */
fun boundsOf(path: List<LatLon>): Bounds {
    require(path.isNotEmpty()) { "Can't compute bounds of an empty path" }
    return Bounds(
        minLat = path.minOf { it.lat },
        minLon = path.minOf { it.lon },
        maxLat = path.maxOf { it.lat },
        maxLon = path.maxOf { it.lon },
    )
}

/** Raw result from a single router call — fetchRoute wraps this into the public RoutedPath (adding `routed = true`). */
data class RouteResult(val path: List<LatLon>, val distanceM: Double, val cues: List<RideCue>)

@Serializable
private data class OsrmResponse(val routes: List<OsrmRoute> = emptyList())

@Serializable
private data class OsrmRoute(
    val distance: Double = 0.0,
    val geometry: OsrmGeometry? = null,
    val legs: List<OsrmLeg> = emptyList(),
)

@Serializable
private data class OsrmGeometry(val coordinates: List<List<Double>> = emptyList())

@Serializable
private data class OsrmLeg(val steps: List<OsrmStep> = emptyList())

@Serializable
private data class OsrmStep(
    val distance: Double = 0.0,
    val name: String? = null,
    val maneuver: OsrmManeuver? = null,
)

@Serializable
private data class OsrmManeuver(val type: String? = null, val modifier: String? = null)

/**
 * Walks an OSRM route's legs/steps into a flat cue list with cumulative
 * distance offsets, so a multi-waypoint route gets one continuous
 * instruction list instead of restarting at zero for every leg.
 */
private fun parseOsrmSteps(route: OsrmRoute): List<RideCue> {
    val cues = mutableListOf<RideCue>()
    var offsetM = 0.0
    for (leg in route.legs) {
        for (step in leg.steps) {
            step.maneuver?.let { maneuver ->
                cues.add(
                    RideCue(
                        atM = offsetM,
                        direction = osrmDirection(maneuver.type.orEmpty(), maneuver.modifier),
                        name = step.name?.trim()?.takeIf { it.isNotEmpty() },
                    ),
                )
            }
            offsetM += step.distance
        }
    }
    return cues
}

/** Routes through every waypoint in order via the public OSRM bike server, capturing turn-by-turn step instructions along the way. */
suspend fun fetchOsrmRoute(points: List<LatLon>): RouteResult {
    val coords = points.joinToString(";") { "${it.lon},${it.lat}" }
    val response = HttpClientProvider.client.get("$OSRM_BIKE/$coords?overview=full&geometries=geojson&steps=true")
    if (!response.status.isSuccess()) error("OSRM ${response.status.value}")
    val route = json.decodeFromString<OsrmResponse>(response.bodyAsText()).routes.firstOrNull() ?: error("OSRM: no route")
    val coordinates = route.geometry?.coordinates.orEmpty()
    if (coordinates.size < 2) error("OSRM: no geometry")
    return RouteResult(path = toLatLonPath(coordinates), distanceM = route.distance, cues = parseOsrmSteps(route))
}

@Serializable
private data class BrouterResponse(val features: List<BrouterFeature> = emptyList())

@Serializable
private data class BrouterFeature(
    val geometry: BrouterGeometry? = null,
    val properties: JsonObject? = null,
)

@Serializable
private data class BrouterGeometry(val coordinates: List<List<Double>> = emptyList())

/**
 * Routes through every waypoint in order via BRouter, respecting the chosen
 * cycling profile. Doesn't return cues: BRouter's public server has no
 * documented, stable step-instruction format to parse (unlike OSRM's
 * steps=true), so a route from here falls back to geometry-detected turns
 * for its cue sheet at nav time instead (Phase 3), same as on web.
 */
suspend fun fetchBrouterRoute(
    points: List<LatLon>,
    profile: BikeProfile,
): RouteResult {
    val lonlats = points.joinToString("|") { "${it.lon},${it.lat}" }
    val url = "$BROUTER_DEFAULT?lonlats=$lonlats&profile=${profile.apiValue}&alternativeidx=0&format=geojson"
    val response = HttpClientProvider.client.get(url)
    if (!response.status.isSuccess()) error("BRouter ${response.status.value}")
    val feature = json.decodeFromString<BrouterResponse>(response.bodyAsText()).features.firstOrNull()
        ?: error("BRouter: no geometry")
    val coordinates = feature.geometry?.coordinates.orEmpty()
    if (coordinates.size < 2) error("BRouter: no geometry")
    val path = toLatLonPath(coordinates)
    // "track-length" comes back as either a JSON string or number depending
    // on the BRouter build; reading it as a raw JsonPrimitive (rather than a
    // typed field) is robust to either — jsonPrimitive.content works for both.
    val declared = feature.properties?.get("track-length")?.jsonPrimitive?.contentOrNull?.toDoubleOrNull()
    return RouteResult(path = path, distanceM = declared ?: pathLengthM(path), cues = emptyList())
}

/**
 * Bike-friendly route through the given waypoints (in order), trying one
 * router then the other. Falls back to straight lines between waypoints
 * when both routers are unreachable.
 *
 * `preferBrouter` puts BRouter first so the requested profile is actually
 * honored — OSRM's public bike profile is fixed and ignores it.
 */
suspend fun fetchRoute(
    points: List<LatLon>,
    profile: BikeProfile = BikeProfile.TREKKING,
    preferBrouter: Boolean = false,
): RoutedPath {
    if (points.size < 2) return RoutedPath(points, pathLengthM(points), routed = false, cues = emptyList())

    val attempts: List<suspend () -> RouteResult> =
        if (preferBrouter) {
            listOf({ fetchBrouterRoute(points, profile) }, { fetchOsrmRoute(points) })
        } else {
            listOf({ fetchOsrmRoute(points) }, { fetchBrouterRoute(points, profile) })
        }

    for (attempt in attempts) {
        try {
            val result = attempt()
            return RoutedPath(result.path, result.distanceM, routed = true, cues = result.cues)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            // try the next router
        }
    }
    return RoutedPath(points, pathLengthM(points), routed = false, cues = emptyList())
}

/**
 * Bike-friendly path from the rider back to the track — port of
 * fetchCyclingRoute in src/lib/rejoin.ts. Deliberately doesn't pass
 * preferBrouter: unlike the planner, speed matters more than profile choice
 * once the rider is already off track, so this takes fetchRoute's default
 * OSRM-first order.
 */
suspend fun fetchCyclingRoute(
    from: LatLon,
    to: LatLon,
): RoutedPath = fetchRoute(listOf(from, to))
